src/Spectre.libGaussianMixtureModelling.Tests/ExpectationMaximizationTest.cpp

```diff
@@ -0,0 +1,287 @@
+#include "BasicInitializationRef.h"
+#include "ExpectationRunnerRef.h"
+#include "MaximizationRunnerRef.h"
+#include "LogLikelihoodCalculator.h"
+
+
+namespace Spectre::libGaussianMixtureModelling
+{
+    class ExpectationMaximizationTest : public ::testing::Test
+    {
+    protected:
+
+        std::vector<double> mzs;
+        std::vector<double> intensities;
+        std::vector<GaussianComponent> gaussianComponents;
+
+        virtual void SetUp() override
+        {
+            gaussianComponents = {
+                { -5.0, 3.0, 0.25 },
+                { 5.0, 3.0, 0.25 },
+                { -2.0, 9.0, 0.5 }
+            };
+
+            mzs = GenerateRange(-20.0, 20.0, 0.5);
+            intensities = GenerateValues(mzs, gaussianComponents);
+        }
+
+        std::vector<double> GenerateRange(double start, double end, double step)
+        {
+            const int size = int((end - start) / step);
+            std::vector<double> result(size);
+            std::fill(result.begin(), result.end(), start);
+
+            for (int i = 0; i < size; i++)
+            {
+                result[i] += step * i;
+            }
+
+            return result;
+        }
+
+        std::vector<double> GenerateValues(std::vector<double> input, 
+            const std::vector<GaussianComponent>& components)
+        {
+            std::vector<double> output(input.size(), 0.0);
+            for (unsigned i = 0; i < input.size(); i++)
+            {
+                for (unsigned j = 0; j < components.size(); j++)
+                {
+                    output[i] += components[j].weight * 
+                        Gaussian(input[i], components[j].mean, components[j].deviation);
+                }
+            }
+            return output;
+        }
+    };
+
+    // Integration test
+    TEST_F(ExpectationMaximizationTest, test_whole_em)
+    {
+        std::mt19937_64 rngEngine(0);
+
+        ExpectationMaximization<
+            BasicInitializationRef,
+            ExpectationRunnerRef,
+            MaximizationRunnerRef,
+            LogLikelihoodCalculator
+        > em(&mzs[0], &intensities[0], (unsigned)mzs.size(), 
+            rngEngine, (unsigned)gaussianComponents.size());
+
+        GaussianMixtureModel result = em.EstimateGmm();
+        std::vector<double> approximates = GenerateValues(mzs, result.components);
+
+        std::vector<double> errors(intensities.size());
+        for (unsigned i = 0; i < intensities.size(); i++)
+        {
+            errors[i] = pow(intensities[i] - approximates[i], 2) * 10000.0; // scale the errors
+        }
+
+        double averageError = 0.0;
+        for (unsigned i = 0; i < errors.size(); i++)
+        {
+            averageError += errors[i];
+        }
+        averageError /= errors.size();
+        double maxError = *std::max_element(errors.begin(), errors.end());
+        double minError = *std::min_element(errors.begin(), errors.end());
+        printf("Avg error: %f\n", averageError);
```

Comment:
Hmmm... [This](https://stackoverflow.com/questions/7923903/printing-additional-output-in-google-test) seems to be too much work for now. Let it be this way.